### ISSUE TITLE
portal: admin dashboard session token renewal strategy

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#2042]
It seems the session token lifetime is not renewed on subsequent authenticated page reloads, so the admin will be unexpectedly logged out as they're using the application.

Just noting here to fix before launch.

cc @AndrewDryga

**Comment by AndrewDryga** [comment]
@jamilbk I was treating session expiration more in a way of "force log in every X hours". Do we want the session to auto-extend though? This will make it possible to stay logged in always as long as you refresh the browser every X hours.

**Comment by jamilbk** [comment]
Hm, perhaps it could work like this:

- Session expiration: 4 hours
- Auto-extend session on liveview navigations and full page loads (user activity)
- **Always** require re-auth for sensitive actions (Deploy Gateway, Add/Update IdP, add users/groups, etc)
- (optional) display session timer in the UI, perhaps the top navbar

I'm hoping to avoid surprise sign outs causing data loss in forms:

1. User onboards, signs in
1. Begins setting up their Idp in Firezone
1. Goes off to their IdP to create a SAML app
1. Comes back to Firezone, enters the needed details in
1. "Create" -> forced reauth -> Form is reset

I suppose we could force the reauth on a strict timer if we cached all form inputs in browser local storage so that they could easily be restored like you did with the original request url, but might be easier to just extend the session

**Comment by jamilbk** [comment]
Cc @rbhitchcock

**Comment by jamilbk** [comment]
Discussion w/ @rbhitchcock this morning (8/4/23) -- 

- Re: UX vs security -- **Bias towards usability**: Session token leaking is easy to prevent -- can use remote IP, user agent chages, etc to flag a re-auth -- time-based methods usually lead to bad UX. Usually only see hard requirement of time-based re-auth periods for the strictest compliance (and these don't even really increase security, and exceptions can be made)
- Will want to have some flexibility in when we force re-auth, and how (e.g. require re-auth, magic link code, MFA for sensitive actions)
- Will eventually want to implement OAuth refresh and access tokens for the end-user auth flow too, so that users aren't forced to manually re-auth their Firezone client every X hours, and we can instead refresh their token on their behalf (post-1.0 feature). Most products _don't_ do this, super annoying to end users (mentioned current 24 hr policy at employer, also our experience at Cisco with the company VPN requiring mandatory 24-hour auth period)

**In summary**: time-based forced re-auth is bad UX, use other methods. Even if compliance requirements appear to require it, in most cases an exception can be made if we document our re-auth logic and how it's just as secure (or better) than strict time-based

**Comment by AndrewDryga** [comment]
@jamilbk 
> Will eventually want to implement OAuth refresh and access tokens for the end-user auth flow too, so that users aren't forced to manually re-auth their Firezone client every X hours, and we can instead refresh their token on their behalf (post-1.0 feature). Most products don't do this, super annoying to end users (mentioned current 24 hr policy at employer, also our experience at Cisco with the company VPN requiring mandatory 24-hour auth period)

For now, we just need to persist cookies for IdP's domain in the WebView, this will guarantee that you don't need to enter username and password as long as IdP will allow you to do so.

> can use remote IP, user agent chages, etc to flag a re-auth

We already do it, only thing is that we will not use IP for the client, because the client connected to VPN can be roaming across cell towers with IP always changing. As for token lifetime, we are using the lifetime provided by IdP and it will be incorrect to increase it (this can lead to some security edge cases that are hard to foresee), but we can definitely have a feature to reduce it for pedantic users.

The admin token will be short-lived though (at most 3 hours) and bound to IP because we simply do not expect people to sit there all day long.

**Comment by jamilbk** [comment]
> IP for the client

Yeah, for clients region is more appropriate, fairly popular request on discovery calls (can be post-1.0)

> it will be incorrect to increase it

Not increase it, but refresh the end-user's access token using the stored refresh token. I realize this only applies to OAuth flows though.

> (at most 3 hours)

3 hours of inactivity though right? We're still planning to refresh from cookie with each request, kinda like [how we had it](https://github.com/firezone/firezone/blob/legacy/apps/fz_http/lib/fz_http_web/auth/html/pipeline.ex#L13) with 0.7?

**Comment by jamilbk** [comment]
I think these might be more reasonable session limits to launch with, until we implement some of the fancier logic above:

Portal sessions: 24 hours
Connlib sessions: 7 days

**Comment by AndrewDryga** [comment]
- [ ] Do not forget to to make sure that auth flow works for service accounts

**Comment by AndrewDryga** [comment]
Questions:
1. Should we extend the session expiration on top of the lifetime of an OIDC ID token returned? 

Pros:
- better usability, no need to log in every X hours with your VPN if you can't configure ID token duration on IdP end

Cons:
- we will allow users to violate company authentication policy / some compliance regulations because we start to m
- not all IdPs return ID a new token as a result of refresh token exchange (Okta does, Google does not)
- "sign out from service" feature on IdP end will not work as expected

Hacks:
- use the refresh token to extend the access token and check if it's still valid (bringing back edge cases when HTTP request failed that we had in 0.7)

**Comment by jamilbk** [comment]
For OIDC, we could also look into implementing a back-channel logout hook that will allow the IdP to log the user upon some event (user terminated, etc): https://openid.net/specs/openid-connect-backchannel-1_0.html#Backchannel

That would reduce the need to couple our own session token lifetime to the IdP's ID token lifetime we're given.

**Comment by AndrewDryga** [comment]
The only thing left after #2696 is to make a recurrent job that will use IdP API to refresh the access token and extend `token.expires_at` every time that happens successfully.

**Comment by jamilbk** [comment]
Maybe should display timer or some other UI notification to admins to let them know the session is about to be cut.
